pwiz_tools/Skyline/FileUI/PeptideSearch/BuildPeptideSearchLibraryControl.cs

```diff
@@ -223,28 +224,139 @@ public bool FilterForDocumentPeptides
+            gridSearchFiles.CellValueChanged += gridSearchFiles_CellValueChanged;
+
+            if (toAdd.Any())
+            {
+                var blibBuild = new BlibBuild(null, toAdd.ToArray());
+                void GetScoreTypes(object sender, DoWorkEventArgs e)
                 {
-                    listSearchFiles.Items.Add(PathEx.RemovePrefix(fileName, dirInputRoot));
+                    IProgressStatus status = new ProgressStatus();
+                    var success = blibBuild.GetScoreTypes(new SilentProgressMonitor(), ref status, out _, out _, out var scoreTypes);
+
+                    // Match input/output files
+                    var filesByLength = toAdd.OrderByDescending(s => s.Length).ToArray();
+                    var scoreTypesTmp = new Dictionary<string, BiblioSpecScoreType[]>();
+                    foreach (var pair in scoreTypes)
+                        scoreTypesTmp[filesByLength.First(f => f.EndsWith(pair.Key))] = pair.Value;
+                    scoreTypes = scoreTypesTmp;
+
+                    Invoke(new MethodInvoker(delegate
+                    {
+                        gridSearchFiles.CellValueChanged -= gridSearchFiles_CellValueChanged;
+
+                        // Gather existing score thresholds
+                        var existingThresholds = new Dictionary<string, double?>();
+                        foreach (var row in gridSearchFiles.Rows.Cast<DataGridViewRow>().Where(row => !toAdd.Contains(FileCellValue.Get(row, columnFile).File)))
+                        {
+                            var scoreType = ScoreTypeCellValue.Get(row, columnScoreType).NameInvariant;
+                            if (scoreType != null)
+                                existingThresholds[scoreType] = ThresholdCell.Get(row, columnThreshold).Threshold;
+                        }
+
+                        for (var i = 0; i < gridSearchFiles.RowCount; i++)
+                        {
+                            var row = gridSearchFiles.Rows[i];
+                            var file = FileCellValue.Get(row, columnFile).File;
+                            if (!toAdd.Contains(file))
+                                continue;
+                            if (success && scoreTypes.TryGetValue(FileCellValue.Get(row, columnFile).File, out var scoreTypesThis))
+                            {
+                                for (var j = 0; j < scoreTypesThis.Length; j++)
+                                {
+                                    var scoreType = scoreTypesThis[j];
+                                    if (j == 0)
+                                    {
+                                        row.Cells[columnScoreType.Index].Value = new ScoreTypeCellValue(scoreType);
+                                    }
+                                    else
+                                    {
+                                        gridSearchFiles.Rows.Insert(i + 1);

```

Comment:
You might want to write a separate function which inserts a row into the grid and sets these three cell values. That function could be used both here and in the place above where a row is being appended to gridSearchFiles.